crypto/x509/v3_purp.c

```diff
@@ -398,30 +417,34 @@ int x509v3_cache_extensions(X509 *x)
     }
     ERR_set_mark();
 
+    /* Cache the SHA1 digest of the cert */
```

Comment:
I'm not convinced that X509_digest() will cache the result if it's not cached yet.

Reply:
> I'm not convinced that X509_digest() will cache the result if it's not cached yet.

Just have a look at the code: it calls `X509_digest()` only in a situation where `x->ex_flags & EXFLAG_SET` is not yet set, and in this case `X509_digest()` does call `asn1_item_digest_with_libctx()`, which will set `x->sha1_hash`.

Reply:
So looking closer, when it is already cached, we do a memcpy() with dest and src the same, which is undefined behaviour, but probably works in practice.

Reply:
The `memcpy()` call is done only when the hash value was already in place, which does not happen when called from `x509v3_cache_extensions()` because this is done only in case `x->ex_flags & EXFLAG_SET` is not true.